Change /oauth/token request specs to use client_secret_basic authentication

ThisIsMissEm opened the pull request:
I noticed this whilst working on the tests for #30329, where we were using `client_secret_post` as the client authentication mechanism, when generally `client_secret_basic` is preferred. This is per OAuth [RFC 6749 Section 2.3.1](https://www.rfc-editor.org/rfc/rfc6749#section-2.3). The names `client_secret_basic` and `client_secret_post` come from [RFC 7591](https://www.rfc-editor.org/rfc/rfc7591.html#section-2)

Here we are indeed using a POST request with body (I assume `params` becomes the body, and not query string params). The rationale for this change is that Doorkeeper's implementation of `client_secret_post` is actually technically broken: https://github.com/doorkeeper-gem/doorkeeper/issues/1769

So this is really playing it on the safe-side by using `client_secret_basic` as the client authentication mechanism.

ThisIsMissEm committed:
Change /oauth/token request specs to use client_secret_basic authentication

ThisIsMissEm commented:
The `Ruby Testing / End to End testing` failures are not at all related to the changes I've made here, as far as I can tell, since these changes are localised to a specific spec.

ThisIsMissEm commented:
@Gargron the end to end testing failures here are apparently related to a change you recently made to the account header?

![image](https://github.com/user-attachments/assets/a5c9b782-7030-4437-81b3-45bde5db2f93)

ThisIsMissEm commented:
@mjankowski I can't reproduce this failure locally at all, and my `.env` file only contains basics:

<img width="726" alt="Screenshot 2025-04-17 at 23 10 57" src="https://github.com/user-attachments/assets/7d1937ee-ada5-485c-a145-aa334dd9e590" />

ThisIsMissEm commented:
I had some failures on the full suite due to those redis URLs, but moving those to `.env.development.local` fixed the failures. Complete suite runs fine on my machine and I cannot reproduce the failures I'm seeing here.

mjankowski commented:
The spec/system JS-enabled failures you saw earlier are almost definitely unrelated to your changes. There are some of those which we see fail intermittently due to a) inherent async nature of JS/browser/server/testing interaction, b) occasionally some issue between the chrome version running on CI and the selenium lib. I suspect in this case it's the former (async JS). These will likely go away either on their own with rebase/push, or by someone on team manually re-running jobs.

The spec request failures, ie - https://github.com/mastodon/mastodon/actions/runs/14523781309/job/40750749656?pr=34474 - do seem related to your changes? I have done zero detail research here, but one guess would be to review the env vars on CI - https://github.com/mastodon/mastodon/blob/main/.github/workflows/test-ruby.yml#L106-L121 - specifically that PAM is enabled (I see some devise pam in the trace), or others? You may have to mirror that full env section to fully reproduce the CI config.

ThisIsMissEm commented:
Well, this change certainly doesn't mess with pam authentication — all it changes is client authentication for mastodon's OAuth API on one endpoint.

ThisIsMissEm commented:
@mjankowski I think this clearly now shows it's not my changes.

mjankowski commented:
Run with something like:

    PAM_ENABLED=true BUNDLE_WITH='pam_authentication test' bin/rspec spec/requests/oauth/token_spec.rb

...will trigger the failure, with backtrace pointing to https://github.com/devkral/devise_pam_authenticatable2/blob/master/lib/devise_pam_authenticatable/strategy.rb#L11

I assume something about having that header present vs not causes an issue here, and/or there's more setup to do for devise to allow this to function with pam enabled.

ThisIsMissEm commented:
I've rebased this against #34501 which I did kind of quickly to test a theory about this HTTP Basic Authentication error we were seeing.

ThisIsMissEm commented:
@mjankowski this is the issue for the underlying issue in devise_pam_authenticatable2: https://github.com/devkral/devise_pam_authenticatable2/issues/3

ThisIsMissEm committed:
Improve /oauth/token request specs

Previously these specs passed incorrect parameters to both the authorization_code
and client_credentials grant flows.

The authorization_code flow does not accept a `scope` parameter, instead the scope
is set when the access grant is created, per RFC 6749 Section 4.1.2. The `code`
parameter is accepted by this flow.

https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2

The client_credentials flow does not accept a `code` parameter, and instead accepts
a `scope` parameter, per RFC 6749 Section 4.4.1

https://www.rfc-editor.org/rfc/rfc6749#section-4.4.1

This ensures we're only testing valid oauth flows, and not deviating from the
specification. The OAuth flows should ignore any unknown parameters (i.e., passing
`code` to client_credentials would have no impact on the functionality, and this
would be asserted at the Doorkeeper level).

ThisIsMissEm commented:
Whilst working on this, I realised that the OAuth Token endpoint specs weren't correct at all: they were passing non-standard parameters to oauth grant flows and testing behavior that could *never* happen (as long as doorkeeper follows the OAuth 2.0 specifications).

For the full details, see the commit message in 8354bb6bf99f70fa388e24e49fd35e3b10c31499

ClearlyClaire commented on spec/requests/oauth/token_spec.rb:
```diff
@@ -1,73 +1,93 @@
 # frozen_string_literal: true
 
 require 'rails_helper'
+require 'debug'
```
I don't think that's intended?

ClearlyClaire commented:
> Whilst working on this, I realised that the OAuth Token endpoint specs weren't correct at all: they were passing non-standard parameters to oauth grant flows and testing behavior that could _never_ happen (as long as doorkeeper follows the OAuth 2.0 specifications).

Yes, there was previously confusion about those in the documentation, and consequently, for some developers. These tests ensure that Mastodon and Doorkeeper do things correctly. I think that testing for proper behavior with bogus parameters makes sense.

ThisIsMissEm commented:
> > Whilst working on this, I realised that the OAuth Token endpoint specs weren't correct at all: they were passing non-standard parameters to oauth grant flows and testing behavior that could _never_ happen (as long as doorkeeper follows the OAuth 2.0 specifications).
> 
> Yes, there was previously confusion about those in the documentation, and consequently, for some developers. These tests ensure that Mastodon and Doorkeeper do things correctly. I think that testing for proper behavior with bogus parameters makes sense.

I guess the question I have here, @ClearlyClaire, is that what exactly are we asserting? Stuff that could've never worked and would have unexpected behavior? The OAuth spec does allow arbitrary parameters to everything too, which is great for extensibility, not so great for "knowing this thing works"

ClearlyClaire commented:
> I guess the question I have here, @ClearlyClaire, is that what exactly are we asserting? Stuff that could've never worked and would have unexpected behavior? The OAuth spec does allow arbitrary parameters to everything too, which is great for extensibility, not so great for "knowing this thing works"

The documentation used to be unclear and there was a lot of confusion as to when to pass scopes, with possible expectation that passing a different set of scopes when redeeming the `authorization_code` to get the token would give you that different set of scopes.

To my knowledge, there has never been a bug that would result in Mastodon treating the `scope` parameter this way at that stage, but we nevertheless wanted to make sure of it.

Back to the original topic of the PR, `POST`ing with `client_id` and `client_secret` has always been the documented way: https://docs.joinmastodon.org/methods/oauth/#token

I'm not against adding a spec for the `client_secret_basic` method, but I would keep the existing spec, as it covers actual use.

ThisIsMissEm committed:
Adjust the specs based on feedback from Claire to ensure we're still testing the invalid scope for authorization code grant flow case

ThisIsMissEm committed:
Ensure the asserted scopes are the ones in the access grant

ThisIsMissEm committed:
Rearrange code to flow better
